File: BOJ/2611.swift
import Foundation

let INF = 987654321
var mn = INF

let n = readLine().map{Int($0)!}!
let m = readLine().map{Int($0)!}!

struct Node {
    let w:Int
    let v:Int
}

var dist = [Int](repeating: 0, count: n+1)
var inDegree = [Int](repeating: 0, count: n+1)
var pre = [Int](repeating: 0, count: n+1)
var adj = [[Node]](repeating: [Node](repeating: Node(w:0,v:0), count: 0), count: n+1)

func topology() {
    var q = [Node]()
    q.append(Node(w: 0, v: 1))
    
    while !q.isEmpty {
        let here = q[0].v
        let cost = q[0].w
        q.removeFirst()
        
        for next in adj[here] {
            let there = next.v
            let nextCost = cost + next.w
            if dist[there] < nextCost {
                dist[there] = nextCost
                pre[there] = here
            }
            inDegree[there] -= 1
            if inDegree[there] == 0 && there != 1 {
                q.append(Node(w: dist[there], v: there))
            }
        }
    }
}

for _ in 0..<m {
    let input = readLine()!.split(separator: " ").map{Int(String($0))!}
    let u = input[0], v = input[1], w = input[2]
    adj[u].append(Node(w:w,v:v))
    inDegree[v] += 1
}

topology()

var res = [1]
var before = pre[1]
while before != 1 {
    res.append(before)
    before = pre[before]
}
res.append(1)
res.reverse()
print(dist[1])
for x in res {
    print(x, terminator: " ")
}

File: BOJ/2611.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10 ** 6)

INF = 987654321

n = int(input())
m = int(input())

adj = [[] for _ in range(n + 1)]
pre = [0] * (n + 1)
dist = [0] * (n + 1)
inDegree = [0] * (n + 1)

for _ in range(m):
    u, v, w = map(int, input().split())
    adj[u].append((w, v))
    inDegree[v] += 1

q = [[0, 1]]

while q:
    here = q[0][1]
    cost = q[0][0]
    q.pop(0)

    for next in adj[here]:
        there = next[1]
        nextCost = cost + next[0]
        if dist[there] < nextCost:
            dist[there] = nextCost
            pre[there] = here
        inDegree[there] -= 1
        if inDegree[there] == 0 and there != 1:
            q.append((dist[there], there))

print(dist[1])
res = [1]
before = pre[1]
while before != 1:
    res.append(before)
    before = pre[before]
res.append(1)

res.reverse()
for x in res:
    print(x, end=" ")
